lib/src/Uncertainty/Algorithm/Simulation/NAIS.cxx

```diff
@@ -0,0 +1,257 @@
+      
+      if (stopCallback_.first && stopCallback_.first(stopCallback_.second))
+        throw InternalException(HERE) << "User stopped simulation";
+    } 
+    
+    // Computation of current quantile
+    currentQuantile = auxiliaryOutputSample.computeQuantile(rhoQuantile_)[0];
+
+    // If failure probability reached, stop the adaptation
+    if (getEvent().getOperator()(currentQuantile, getEvent().getThreshold()))
+    {
+      currentQuantile = getEvent().getThreshold();
+    }
+    else
+    {
+      // Computation of weights
+      weights = computeWeights(auxiliaryInputSample, auxiliaryOutputSample, currentQuantile, auxiliaryDistribution);
+
+      // Update of auxiliary distribution
+      auxiliaryDistribution = computeAuxiliaryDistribution(auxiliaryInputSample, weights);
+    }
+    
+    if (stopCallback_.first && stopCallback_.first(stopCallback_.second))
+      throw InternalException(HERE) << "User stopped simulation";
+      
+  } // while
+
+  // Find failure sample indices
+  Indices indicesCritic(0);
+  for (UnsignedInteger i = 0; i < auxiliaryOutputSample.getSize(); ++i)
+  {
+    // Find failure Points
+    if (getEvent().getOperator()(auxiliaryOutputSample(i, 0), getEvent().getThreshold()))
+      indicesCritic.add(i);
+  } // for i
+
+  const Sample resp_sampleCritic(auxiliaryOutputSample.select(indicesCritic));
+  const Sample sampleCritic(auxiliaryInputSample.select(indicesCritic));
+
+  // Evaluate initial log PDF in parallel on failure sample
+  Sample logPDFInitCritic(initialDistribution_.computeLogPDF(sampleCritic));
+
+  // Evaluate auxiliary log PDF in parallel on failure sample
+  Sample logPDFAuxiliaryCritic(auxiliaryDistribution.computeLogPDF(sampleCritic));
+
+  Scalar sumPdfCritic = 0.0;
+  for(UnsignedInteger i = 0; i < indicesCritic.getSize(); ++i)
+  {
+    sumPdfCritic += std::exp(logPDFInitCritic(i, 0) - logPDFAuxiliaryCritic(i, 0));
+  }
+  Scalar failureProbability = sumPdfCritic / numberOfSample;
+  Scalar varianceCritic = 0.0;
+  Scalar varianceCriticTemporary;
+  for(UnsignedInteger i = 0; i < indicesCritic.getSize(); ++i)
+  {
+    varianceCriticTemporary = std::exp(logPDFInitCritic(i, 0) - logPDFAuxiliaryCritic(i, 0))-failureProbability;
+    varianceCritic += varianceCriticTemporary*varianceCriticTemporary;
+  }  // for i 
+  
+  Scalar variancenonCritic = (numberOfSample -indicesCritic.getSize()) * (failureProbability*failureProbability);
+  Scalar varianceEstimate = (varianceCritic + variancenonCritic)/(numberOfSample-1)/numberOfSample ;
+
+  // Save of data in Simulation naisResult_ structure
+  naisResult_.setProbabilityEstimate(sumPdfCritic / numberOfSample);
+  naisResult_.setAuxiliaryDistribution(auxiliaryDistribution);
+  naisResult_.setAuxiliaryInputSample(auxiliaryInputSample);
+  naisResult_.setAuxiliaryOutputSample(auxiliaryOutputSample);
+  naisResult_.setWeights(weights);
+  naisResult_.setOuterSampling(getMaximumOuterSampling());
```

Comment:
here you want the total number of blocks, so getMaximumOuterSampling() * number of iterations

Reply:
for "number of iterations" you means number of while loop in the run()?